### PR TITLE
Federation: Minor corrections in service controller and add a unit testcase

### PR DESCRIPTION
**Pull request by shashidharatd**
**What this PR does / why we need it**:
This PR fixes few outdated comments in federation service controller and few other minor fixes.
This also adds a unit test case to test federated service deletion.

/assign @quinton-hoole 
/cc @marun @kubernetes/sig-federation-pr-reviews 

```release-note
NONE
```

### CODE REVIEW

File: federation/pkg/federation-controller/service/servicecontroller.go [hunk 1]
```diff
@@ -68,16 +69,13 @@ var (
 )
 
 type ServiceController struct {
-	federationClient fedclientset.Interface
-	// A store of services, populated by the serviceController
-	serviceStore corelisters.ServiceLister
-	// Watches changes to all services
+	federationClient  fedclientset.Interface
+	serviceStore      corelisters.ServiceLister
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/servicecontroller.go]: Why remove the comments?

File: federation/pkg/federation-controller/service/servicecontroller.go [hunk 2]
```diff
@@ -68,16 +69,13 @@ var (
...
 	serviceController cache.Controller
 	federatedInformer fedutil.FederatedInformer
 	eventBroadcaster  record.EventBroadcaster
 	eventRecorder     record.EventRecorder
-	// services that need to be synced
-	queue *workqueue.Type
+	queue             *workqueue.Type
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/servicecontroller.go]: As above.

File: federation/pkg/federation-controller/service/servicecontroller.go [hunk 3]
```diff
@@ -179,10 +176,6 @@ func New(federationClient fedclientset.Interface) *ServiceController {
 			svc := obj.(*v1.Service)
 			orphanDependents := false
 			err := client.Core().Services(svc.Namespace).Delete(svc.Name, &metav1.DeleteOptions{OrphanDependents: &orphanDependents})
-			// IsNotFound error is fine since that means the object is deleted already.
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/servicecontroller.go]: Why was this removed?

File: federation/pkg/federation-controller/service/servicecontroller.go [hunk 4]
```diff
@@ -235,44 +228,40 @@ func (s *ServiceController) updateService(obj pkgruntime.Object) (pkgruntime.Obj
 	return s.federationClient.Core().Services(service.Namespace).Update(service)
 }
 
-// Run starts a background goroutine that watches for changes to federation services
-// and ensures that they have Kubernetes services created, updated or deleted appropriately.
-// federationSyncPeriod controls how often we check the federation's services to
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/servicecontroller.go]: Why remove these comments?

File: federation/pkg/federation-controller/service/servicecontroller.go [hunk 5]
```diff
@@ -235,44 +228,40 @@ func (s *ServiceController) updateService(obj pkgruntime.Object) (pkgruntime.Obj
...
-// (e.g. directly via the underlying cluster's API) are correctly accounted for.
-
-// It's an error to call Run() more than once for a given ServiceController
-// object.
+// Run starts informers, delay deliverers and workers. Workers continuously watch for events which could
+// be from federation or federated clusters and tries to reconcile the service objects from federation to
+// federated clusters.
 func (s *ServiceController) Run(workers int, stopCh <-chan struct{}) {
 	glog.Infof("Starting federation service controller")
+	defer glog.Infof("Shutting down federation service controller")
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/servicecontroller.go]: Deferred statements are executed in LIFO order, so you will log "Shutting down" after you have already shut down.  I don't think that's what you want.

**Comment by quinton-hoole** [comment]
More review still to come.  It's late :)

File: federation/pkg/federation-controller/service/servicecontroller.go (hunk 1 above)
Comment (shashidharatd) [inline federation/pkg/federation-controller/service/servicecontroller.go]: i thought they are self-explanatory.
we use `federationClient` not only for watching changes to service, but also to update them when required.

File: federation/pkg/federation-controller/service/servicecontroller.go (hunk 2 above)
Comment (shashidharatd) [inline federation/pkg/federation-controller/service/servicecontroller.go]: yeah this one makes sense, adding it back

File: federation/pkg/federation-controller/service/servicecontroller.go (hunk 3 above)
Comment (shashidharatd) [inline federation/pkg/federation-controller/service/servicecontroller.go]: It is already taken care of here. https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/util/federated_updater.go#L125

File: federation/pkg/federation-controller/service/servicecontroller.go (hunk 4 above)
Comment (shashidharatd) [inline federation/pkg/federation-controller/service/servicecontroller.go]: it is rephrased to below.
```
+// Run starts informers, delay deliverers and workers. Workers continuously watch for events which could
+// be from federation or federated clusters and tries to reconcile the service objects from federation to
+// federated clusters.
```

File: federation/pkg/federation-controller/service/servicecontroller.go (hunk 5 above)
Comment (shashidharatd) [inline federation/pkg/federation-controller/service/servicecontroller.go]: yes agree, that was unintentional. will fix that

**Comment by shashidharatd** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by shashidharatd** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by marun** [comment]
I don't find anything controversial here.  Looking forward to a service controller based on the sync controller in 1.8 to avoid duplicated test effort (like the addition for deletion).

/lgtm 

---

Reviewed 1 of 4 files at r1.
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[federation/pkg/federation-controller/service/servicecontroller.go, line 100 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/46483#-KlVUcDYVdWOAUvtKlHr:-KlVUcDYVdWOAUvtKlHs:b-z3upxw) ([raw file](https://github.com/kubernetes/kubernetes/blob/6e66a8860fb4ee85ae5eff8afbc3fa05e2f047b0/federation/pkg/federation-controller/service/servicecontroller.go#L100)):*
> ```Go
> 
> // New returns a new service controller to keep service objects between
> // federation and federated clusters in sync.
> ```

(No action required) 

s/federation/the federation control plane/ ? 
s/federated clusters/member clusters/ ?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/46483#-:-KlVVNmggTa-turUCsSg:bl0v2o3)*

**Comment by shashidharatd** [comment]
Thanks for lgtm @marun. Agree with you. we shall migrate service controller to extend based on sync controller early in 1.8. 

---

Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[federation/pkg/federation-controller/service/servicecontroller.go, line 100 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/46483#-KlVUcDYVdWOAUvtKlHr:-KlW6XT3MHg06PQBGlMP:b-xqfvjd) ([raw file](https://github.com/kubernetes/kubernetes/blob/6e66a8860fb4ee85ae5eff8afbc3fa05e2f047b0/federation/pkg/federation-controller/service/servicecontroller.go#L100)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) 

s/federation/the federation control plane/ ? 
s/federated clusters/member clusters/ ?
</blockquote></details>

Handled.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/46483#-:-KlW6ZZ_E7Wp2E_U1SuA:b6sxhnw)*

File: federation/pkg/federation-controller/service/servicecontroller.go (hunk 1 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/servicecontroller.go]: Thanks.  Makes sense.  Sorry, I misread the code.

File: federation/pkg/federation-controller/service/servicecontroller.go (hunk 3 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/servicecontroller.go]: Thanks, makes sense.

**Comment by shashidharatd** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
